Add data project template (#17)
snowflakedb/snowflake-cli-templates: data_project/README.md

# Data Project template

## Introduction

> This is the template which can be considered as a starter of the Data Project in Snowflake. The Data
> Project allows applying changes to the account infrastructure in a declarative way which make this
> process more convenient comparing it to the imperative way. Data Projects automatically detect changes
> that should be applied to users infrastructure (creates, alters and drops) and take care of execution
> of defined statements in the correct order on its own.

Reviewer: ```suggestion
This template can be used as a starting point for a Snowflake Data Project. 

Data Projects enable declarative infrastructure management, presenting a less error-prone paradigm that integrates better with Git than imperative approaches. Through automated change detection, Data Projects keep the Snowflake environment in-sync with the definitions files in version management. New objects will be created, existing objects modified, and no longer needed objects deleted. This enables the use of CI/CD best practices when working with Snowflake environments.
```

## Directory Structure

```
<project_name>
      ├── definitions (2)
      │   ├────── main.sql (3)
      │   └────── [...]
      ├── manifest.yml (1)
      ├── snowflake.yml (4)
      └── template.yml (5)
```

1. [manifest.yml][manifest] - this is the file where you can:

>     * specify what files should be included in the version creation process. In the `include_definitions` list you can:
>       * attach a specific file by passing its relative path to the root of the project, e.g. `definitions/main.sql`
>       * attach all files from a specific directory, e.g. `definitions/.*`
>     * specify values for template variables. Definition files (templates) are rendered and filled up with variable values during the project execution process. Variable values are defined in the `template_variables` list, e.g. `example_db_name: "db1"`

Reviewer: ```suggestion
    * specify what files should be included in the version creation process. In the `include_definitions` list you can:
      * include a specific file by passing its relative path to the root of the project, e.g. `definitions/main.sql`
      * include all files from a specific directory, e.g. `definitions/.*`
    * specify values for template variables. Definition files (templates) are rendered as Jinja2 variables during the project execution process. Variable values are defined in the `template_variables` list, e.g. `example_db_name: "db1"`
```

> 2. `definitions` - this is the directory that is initially configured in the [manifest.yml][manifest] to be a place for all `.sql` files containing project entities definitions. You can also arrange the directory/file structure by yourself.

Reviewer: ```suggestion
2. `definitions` - this is the default directory as defined in the [manifest.yml][manifest] for all `.sql` files containing project entity definitions. You are free to use an arbitrarily nested directory structure.
```

> 3. [main.sql][main.sql] - this is the file that contains some example definitions of project entities. You define particular entities with a `DEFINE` keyword which behaves similar to `CREATE OR ALTER`, e.g. `DEFINE DATABASE d1 COMMENT = 'some comment'`. You can also replace [main.sql][main.sql] or create more `.sql` files to organise the code better.

Reviewer: ```suggestion
3. [main.sql][main.sql] - this is the file that contains some example definitions of project entities. You define particular entities with a `DEFINE` keyword which behaves similar to `CREATE OR ALTER`, e.g. `DEFINE DATABASE d1 COMMENT = 'some comment'`. Removing a `DEFINE` statement results in the entity being dropped. You can also replace [main.sql][main.sql] or create more `.sql` files to organise the code better.
```

4. [snowflake.yml][snowflake] - this is the file required by the `Snowflake CLI`. The most important keys are:
    * `identifier` - specifies the name of `PROJECT` entity managed in Snowflake.
    * `stage` - specifies the name of `STAGE` entity that stores project files in Snowflake.

>     * `main_file` - the path to the projects' manifest file.

Reviewer: Please verify if this property is actually being used. I am not sure.

Author: It looks like this property doesn't have any functional usage, but when it's missing, it results in an error in the `snow project` commands.

>     * `artifacts` - the list of paths to files/directories that should be used by the `Snowflake CLI` in Data Project management.

Reviewer: ```suggestion
    * `artifacts` - the list of files and directories that make up the Snowflake Data Project. The Snowflake CLI will upload them to the stage when creating new project versions.
```

5. [template.yml][template] - this file is used by the `Snowflake CLI` to generate a new project from this template.

## Working on Data Projects with Snowflake CLI

### 1. Initialize Data Project from the template

> In order to initialize a new Data Project from this template execute the command below and provide data required in command prompts. This command will create a new directory with Data Project files. Please fill `<project_dir_name>` with your own name.

Reviewer: ```suggestion
In order to initialize a new Data Project from this template execute the command below and provide data required in command prompts. This command will create a new directory with Data Project files. Please fill `<project_dir_name>` with where you want the project directory to be created.
```

```

example usage:

```bash
snow init MY_PROJECT --template data_project
```

### 2. Define entities in `.sql` files

In this step you need to define entities that will be managed by the project. You can define these
entities in the prepared [definitions/main.sql][main.sql] file, but you can also create your own files. If you
decide to add more `.sql` files, make sure that they will be included in the process of creating a new
project version by adding their paths to `include_definitions` list in the [manifest.yaml][manifest] file.

An example content of the definition file:
```sql
    DEFINE ROLE role1;
    DEFINE DATABASE ROLE role2;

    DEFINE DATABASE db1;
    DEFINE DATABASE db2;

    DEFINE SCHEMA db1.sch1;
    DEFINE SCHEMA db2.sch2;

    DEFINE TABLE db1.sch1.tb1 (col_1 integer, col_2 varchar);
    DEFINE TABLE db2.sch2.tb2 (col_3 integer, col_4 varchar);
```

### 3. Create Project

Once the entity definitions included in definition files are ready to be applied to your infrastructure,
it's required to create a `PROJECT` entity with a new `VERSION` from your local files. You can perform this
operation by executing the command below:

```bash

> snow project create-version

Reviewer: This is not the correct command for `create project`

Author: It is the correct command. There is no `create project` command. I discussed it with @sfc-gh-anugupta today and we would like this command to be renamed to `create` only, so the syntax will be `snow project create`.

<img width="1713" alt="image" src="https://github.com/user-attachments/assets/7581fa24-c33e-4ed5-bf0f-1a6949180f4e" />

Reviewer: I believe the `create project` is executed under the hood by `snow project add-version`. If not, there is also `snow object create` which should work. https://docs.snowflake.com/en/developer-guide/snowflake-cli/command-reference/object-commands/create

Author: No, when you execute `snow project add-version`, the PROJECT should exist at the time. This command doesn't create it. The project is created (and the version is added) in the `snow project create-version`.

Reviewer 2: snowflake.yml - type: project -> type does not exist in yaml syntax.

snow project create-version -> THIS RUNS "CREATE PROJECT IF NOT EXISTS IDENTIFIER('<project_name>')" against WH, this is not publicly ready at all.

│ 001003 (42000): 01bb01a8-0001-6f0a-0001-6bfa001f303e: SQL compilation error:                                                                                                                          │
│ syntax error line 1 at position 7 unexpected 'PROJECT'.                                                                                                                                               │
│ syntax error line 1 at position 18 unexpected 'NOT'.                                                                                                                                                  │
│ syntax error line 1 at position 39 unexpected '('.

```

This command will create a new `STAGE` if it doesn't already exist or use an existing one as a target
of local files deployment, from which, the new `VERSION` will be created. The `STAGE` and the `PROJECT`
will be created in the current sessions' database and schema or in these, which are specified in the
flags of `snow` command. Name of the `PROJECT` is specified in the [snowflake.yml][snowflake] file under the `identifier`
key and `STAGE` name is specified under the `stage` key.

You can also use this command for adding new versions, not only the initial one.

### 4. Project dry-run [optional]

> Before the real execution of the PROJECT, you can validate what changes will be applied to your infrastructure.
> In purpose of this action, you can perform the dry-run execution with the command below. This command won't
> persist any changes in the infrastructure.

Reviewer: ```suggestion
After creating a new PROJECT version, you can validate what changes will be applied to your Snowflake account with this command. This command will perform all the same validations and consistency checks like a regular `snow project execute`, but will not persist any changes to your Snowflake objects.
```

```bash
snow project dry-run <project_name> --version <version_name>
```

example usage:

```bash
snow project dry-run <! name | to_snowflake_identifier !> --version latest
```

### 5. Execute Project

> In order to apply changes to your infrastructure you need to execute the particular version of the PROJECT.

Reviewer: ```suggestion
In order to apply changes to your Snowflake account you need to execute the particular version of the PROJECT. It is recommended to first perform a dry-run with the changes.
```

You can do this with the following command:

```bash

> snow project execute <project_name> --version <version_name>

Reviewer 3: Can `project_name` be templated?

# Data Project template

## Introduction

This template can be used as a starting point for a Snowflake Data Project.

> Data Projects enable declarative infrastructure management, presenting a less error-prone paradigm

Author: Honestly, I think that our target will understand this introduction. I can also explain what imperative and declarative infrastructure management is, but it will be too general and I assume that our recipients (mostly dev-ops or data engineers) will already know this topic.

If you could elaborate more on which phrases make you confused, maybe we can reach a compromise between being too general and being to technically specific.

that integrates better with Git than imperative approaches. Through automated change detection, Data
Projects keep the Snowflake environment in-sync with the definitions files in version management.
New objects will be created, existing objects modified, and no longer needed objects deleted.
This enables the use of CI/CD best practices when working with Snowflake environments.

## Directory Structure

```
<project_name>
      ├── definitions (2)
      │   ├────── main.sql (3)
      │   └────── [...]
      ├── manifest.yml (1)
      ├── snowflake.yml (4)
      └── template.yml (5)
```

> 

Author: These numbers refer to the numbers that stand by file/dir names in the files tree graph above. That's right; order doesn't matter, but it's some kind of navigation.

Regarding <project_name>, I provided a description of this directory.

> 1. [manifest.yml][manifest] - this is the file where you can:

Author: I applied most of the recommended content.
